src/module.c

```diff
@@ -7389,6 +7415,225 @@ RedisModuleBlockedClient *moduleBlockClient(RedisModuleCtx *ctx, RedisModuleCmdF
+void moduleInvokeFreePrivDataCallback(client *c, RedisModuleBlockedClient *bc) {
+    if (bc->privdata && bc->free_privdata) {
+        RedisModuleCtx ctx;
+        int ctx_flags = c == NULL ? REDISMODULE_CTX_BLOCKED_DISCONNECTED : REDISMODULE_CTX_NONE;
+        moduleCreateContext(&ctx, bc->module, ctx_flags);
+        ctx.blocked_privdata = bc->privdata;
+        ctx.client = bc->client;
+        bc->free_privdata(&ctx,bc->privdata);
+        moduleFreeContext(&ctx);
+    }
+}
+
+/* Returns 1 if the module has any custom auth callbacks registered. If `should_unregister` is
+ * non-zero, all the module's custom auth cbs are unregistered.
+ * Returns 0 otherwise - if the Module has no custom auth cb registered. */
+int moduleSearchOrUnregisterCustomAuthCBs(RedisModule *module, int should_unregister) {
+    int callback_found = 0;
+    listIter li;
+    listNode *ln;
+    listRewind(moduleCustomAuthCallbacks, &li);
+    while ((ln = listNext(&li))) {
+        RedisModuleCustomAuthCtx *ctx = listNodeValue(ln);
+        if (ctx->module == module) {
+            callback_found = 1;
+            if (!should_unregister) return callback_found;
+            listDelNode(moduleCustomAuthCallbacks, ln);
+            zfree(ctx);
+        }
+    }
+    return callback_found;
+}
+
+/* Returns 1 if custom module based auth is in progress on any client. Returns 0 otherwise. */
+int isCustomAuthInProgress() {
+    if (!listLength(moduleCustomAuthCallbacks)) return 0;
+    listIter li;
+    listNode *ln;
+    listRewind(server.clients,&li);
+    while ((ln = listNext(&li)) != NULL) {
+        client *c = listNodeValue(ln);
+        if (clientHasModuleAuthInProgress(c)) {
+            return 1;
+        }
+    }
+    return 0;
+}
+
+/* Search for & attempt next custom auth callback after skipping the ones already attempted.
+ * Returns the result of the custom auth callback. */
+int attemptNextCustomAuthCb(client *c, robj *username, robj *password, const char **err) {
+    int skipped_prev_callbacks = c->custom_auth_ctx == NULL;
+    RedisModuleCustomAuthCtx *cur_auth_ctx = NULL;
+    listNode *ln;
+    listIter li;
+    listRewind(moduleCustomAuthCallbacks, &li);
+    int result = REDISMODULE_AUTH_NOT_HANDLED;
+    while((ln = listNext(&li))) {
+        cur_auth_ctx = listNodeValue(ln);
+        if (skipped_prev_callbacks) {
+            /* Remove the custom auth complete flag before we attempt the next cb. */
+            c->flags &= ~CLIENT_CUSTOM_AUTH_HAS_RESULT;
+            RedisModuleCtx ctx;
+            moduleCreateContext(&ctx, cur_auth_ctx->module, REDISMODULE_CTX_NONE);
+            ctx.client = c;
+            *err = NULL;
+            RedisModuleString *err_string = NULL;
+            c->custom_auth_ctx = cur_auth_ctx;
+            result = cur_auth_ctx->auth_cb(&ctx, username, password, &err_string);
+            if (err_string) {
+                *err = err_string->ptr;
+                decrRefCount(err_string);
```

Comment:
This will not work since the string might get freed in decrRefCount, which will make it unavailable. In the error reply.